pkg/main/router.go

```diff
@@ -0,0 +1,131 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"strings"
+	"sync"
+)
+
+// Router an interface for delegating function invocations to idle servers
+type Router interface {
+	Delegate(input map[string]interface{}) (*Response, error)
+}
+
+// DefaultRouter a struct that hold servers that can be delegated to
+type DefaultRouter struct {
+	servers []Server
+	mutex   *sync.Mutex
+}
+
+// NewRouter constructor for DefaultRouters
+func NewRouter(servers []Server) *DefaultRouter {
+	return &DefaultRouter{
+		servers: servers,
+		mutex:   &sync.Mutex{},
+	}
+}
+
+// Delegate delegates function invocation to an idle server
+func (r *DefaultRouter) Delegate(input map[string]interface{}) (*Response, error) {
+	server, _ := r.findFreeServer()
+	defer r.releaseServer(server)
+
+	var e Error
+	resp, err := server.Invoke(input)
+	if resp != nil {
+		defer resp.Close()
+	}
+	if err != nil {
+		if _, ok := err.(TimeoutError); ok {
+			e = Error{
+				ErrorType: FunctionError,
+				Message:   "Invocation exceeded the timeout",
+			}
+		} else if _, ok := err.(BadRequestError); ok {
+			e = Error{
+				ErrorType: InputError,
+				Message:   "Input invalid",
+			}
+		} else if _, ok := err.(InvocationError); ok {
+			e = Error{
+				ErrorType: FunctionError,
+				Message:   "Failed invoking function.",
+			}
+		} else if _, ok := err.(InvalidResponsePayloadError); ok {
+			e = Error{
+				ErrorType: InputError,
+				Message:   "Failed serialzing function return return",
+			}
+		} else if _, ok := err.(UnknownSystemError); ok {
+			e = Error{
+				ErrorType: SystemError,
+				Message:   "An unknown error occured",
+			}
+		} else {
+			e = Error{
+				ErrorType: SystemError,
+				Message:   "Something went wrong.",
+			}
+		}
+	}
+
+	logs := Logs{
+		Stdout: splitLogsOnNewline(server.Stdout()),
```

Comment:
I'm not sure yet, but this might be a race condition. There is a separate goroutine that reads the command's output over the pipe and copies it to the buffer, so it could be that the server returned a response but the buffer has not been filled up yet, and it seems that the only way to ensure copying has been finished is calling Wait on command (which won't work in this case). You might need to use `os.Pipe()` instead.

Reply:
If you look at the first commit, I did try using the os.Pipe() for stderr and stdout but ran into a bunch of issues trying to read from the ReadCloser. One of the biggest issues I had was that the call to Read() would block if there was nothing in stdout/stderr.

Reply:
Ah, got it. Do we expect every function call to produce something on stderr/stdout on function server? If so, you could spin a goroutine to read from the pipe and use a channel to synchronize. When using `Buffer`, `os.Command` uses pipe under the hood anyway (https://github.com/golang/go/blob/go1.10/src/os/exec/exec.go#L267), It just spins a goroutine itself to copy data. since it uses goroutine, reading from that buffer means a race condition - although I'm guessing due to a delay of TCP stack it may never be observed. 
But if we can't assume any data on stderr/stdout, then I guess we can't do anything else...

Reply:
Even if we can guarantee something on stderr/stdout we would still run into the problem that we don't know when we have exhausted the contents of the pipe without doing a final Read that will ultimately block until more input is received or the pipe is closed. This would be a non-issue if the ReadCloser would just return (0, nil) instead of blocking on reads when the pipe is 'empty'